ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala

```diff
@@ -0,0 +1,191 @@
+      // T1 Attempts to add user1 and fails with a constrain violation exception.
+      // Expected: Fail with UserNotFound exception or prevent the failure from occurring.
+      if (userExists(user.id, connection)) {
+        Left(UserExists(userId = user.id))
+      } else {
+        backend.createUser(user)(connection)
+        rights.foreach(right => backend.addUserRight(id = user.id, right = right)(connection))
+        Right(())
+      }
+    }.map(_.map { res =>
+      logger.info(s"Created new user: ${user} with rights: ${rights.mkString(", ")}")
+      res
+    })(scala.concurrent.ExecutionContext.parasitic)
+  }
+
+  override def deleteUser(id: UserId): Future[Result[Unit]] = {
+    inTransaction { connection =>
+      if (!userExists(id, connection)) {
+        Left(UserNotFound(userId = id))
+      } else {
+        Right(backend.deleteUser(id = id)(connection = connection))
+      }
+    }.map(_.map { res =>
+      logger.info(s"Deleted user with id: ${id}")
+      res
+    })(scala.concurrent.ExecutionContext.parasitic)
+  }
+
+  override def getUser(id: UserId): Future[Result[domain.User]] = {
+    inTransaction { connection =>
+      backend
+        .getUser(id = id)(connection)
+        .fold[Either[UserManagementStore.Error, domain.User]](
+          Left(UserManagementStore.UserNotFound(userId = id))
+        )(
+          Right(_)
+        )
+    }
+  }
+
+  override def grantRights(
+      id: UserId,
+      rights: Set[domain.UserRight],
+  ): Future[Result[Set[domain.UserRight]]] = {
+    inTransaction { connection =>
+      // For PostgreSQL on READ_COMMITTED isolation level the following undesirable results can occur:
+      // 1)
+      // T1 Adds some rights to user1 but haven't committed yet.
+      // T2 Deletes user1 and commits.
+      // T1 Attempts to add some more rights to user1 and fails on foreign key constraint violation.
+      // Expected: Fail with UserNotFound exception or prevent the failure from occurring.
+      //
+      // 2)
+      // T1 Checks if user-right1 exists for user1.
+      // T2 Adds user-right1 to user user1 and commits.
+      // T1 Attempts to add user-right1 to user1 and fails on a unique constraint violation.
+      // Expected: Requesting to grant a right that a user already posses should always succeed.
+      if (!userExists(id, connection)) {
+        Left(UserNotFound(id))
+      } else {
+        val addedRights = rights.filter { right =>
+          if (!backend.userRightExists(id = id, right = right)(connection)) {
+            backend.addUserRight(id = id, right = right)(connection)
+            true
+          } else {
+            false
+          }
+        }
+        Right(addedRights)
+      }
+    }.map(_.map { res =>
+      if (res.nonEmpty)
+        logger.info(
+          s"Granted user rights to user id: ${id}. Granted rights: ${rights.mkString(", ")}"
+        )
+      res
+    })(scala.concurrent.ExecutionContext.parasitic)
+  }
+
+  override def revokeRights(
```

Comment:
same pattern as with the grant: room for improvement.
also would consider doing something similar in InMemoryUserManagementStore: withUser helper